src/class-model.ts

```diff
@@ -274,10 +270,24 @@ export function register<Instance, Klass extends { new (...args: any[]): Instanc
   klass.volatiles = mstVolatiles;
 
   // conform to the API that the other MQT types expect for creating instances
-  klass.instantiate = (snapshot, context) => new klass(snapshot, context.env, context);
+  klass.instantiate = (snapshot, context, parent) => new klass(snapshot, context, parent);
   (klass as any).is = (value: any) => value instanceof klass || klass.mstType.is(value);
   klass.create = (snapshot, env) => klass.mstType.create(snapshot, env);
-  klass.createReadOnly = (snapshot, env) => new klass(snapshot, env) as any;
+  klass.createReadOnly = (snapshot, env) => {
+    const context: InstantiateContext = {
+      referenceCache: new Map(),
+      referencesToResolve: [],
+      env,
+    };
+
+    const instance = new klass(snapshot, context, null) as any;
+
+    for (const resolver of context.referencesToResolve) {
+      resolver();
+    }
```

Comment:
We used to only do this `if (isRoot)`.

Can this be done everytime because nested objects will be created via `instantiate`?

Reply:
Sorta -- it used to be that the constructor played double duty as the entry point for new root level instances and new child instances. To avoid checking rootness all the time, I made the constructor only do the always stuff, and moved the stuff that only happens for the root to the one place that creates roots (`.createReadOnly`)